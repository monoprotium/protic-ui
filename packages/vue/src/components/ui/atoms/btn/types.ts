  export enum PrtBtnSize {
    sm = 'sm',
    base = 'base',
    lg = 'lg',
  }

  export enum PrtBtnVariant {
    solid = 'solid',
    outline = 'outline',
    ghost = 'ghost',
  }
  

