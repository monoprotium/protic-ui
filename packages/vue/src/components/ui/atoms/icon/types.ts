export enum PrtIconSize {
    xs = '10px',
    sm = '12px',
    base = '16px',
    lg = '24px',
    xl = '30px',
    '2xl' = '36px',
    '3xl' = '42px',
    '4xl' = '48px',
}
  
