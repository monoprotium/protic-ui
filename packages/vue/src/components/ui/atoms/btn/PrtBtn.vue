<script setup lang="ts">
import {  type ConcreteComponent, computed, toRefs,  withDefaults } from 'vue';
import { PrtBtnSize, PrtBtnVariant } from './types';



const props = withDefaults(defineProps<{
  size: `${PrtBtnSize}`;
  variant: `${PrtBtnVariant}`;
  disabled: boolean;
  square: boolean;
  tag: string | ConcreteComponent;
  loading: boolean;
  fullWidth: boolean;
  color: string;
}>(), {
  size: PrtBtnSize.base,
  variant: PrtBtnVariant.solid,
  disabled: false,
  square: false,
  tag: 'button',
  loading: false,
  fullWidth: false,
  color: 'bg-el-7',
});

const { size, square, tag, variant, fullWidth, loading, disabled, color } = toRefs(props);

const sizeClass = computed(() => {
  const base = 'text-center';
  const padding = square.value ? 'p-2' : 'py-2 px-4';
  const minWidth = square.value ? 'min-w-10.5' : '';
  const textSize = size.value === PrtBtnSize.sm ? 'text-sm' : 'text-base';
  const paddingSize = {
    [PrtBtnSize.base]: padding,
    [PrtBtnSize.sm]: square.value ? 'p-1.5 min-w-8.5' : 'py-1.5 px-3',
    [PrtBtnSize.lg]: square.value ? 'p-4 min-w-14.5' : 'py-3 px-6',
  };
  return `inline-flex items-center justify-center font-medium transition-colors duration-150 rounded ${base} ${minWidth} ${textSize} ${paddingSize[size.value]}`;
});

const variantClass = computed(() => {
  const variantColor = color.value.replace('bg-', '');
  const variantColorOpposite = "white";
  const isDisabled = disabled.value;
  const textColor = isDisabled ? 'text-gray-400' : 'text-white';


  const backgroundColor = isDisabled ? 'bg-gray-500' : color.value;
  const borderColor = `border-${variantColor}`;
  const focusStyles = `focus:ring-2 focus:ring-offset-2 focus:ring-${variantColor} focus:ring-opacity-50`;

  const styleMap = {
    [PrtBtnVariant.solid]: `${textColor} ${backgroundColor} hover:bg-opacity-90 ${focusStyles}`,
    [PrtBtnVariant.outline]: `bg-transparent ${borderColor} hover:text-${variantColorOpposite} text-${variantColor} hover:bg-${variantColor} ${focusStyles}`,
    [PrtBtnVariant.ghost]: `text-${variantColor} bg-transparent hover:bg-${variantColor} hover:text-${variantColorOpposite} ${focusStyles}`,
  };

  return styleMap[variant.value] || '';
});

const buttonClasses = computed(() => {
  const baseClasses = `${sizeClass.value} ${fullWidth.value ? 'w-full' : ''} ${disabled.value ? 'opacity-50 cursor-not-allowed' : ''} ${loading.value ? 'cursor-wait' : ''}`;
  return `${baseClasses} ${variantClass.value}`.trim();
});

const componentTag = computed(() => tag.value || 'button');
</script>

<template>
  <component :is="componentTag" :type="componentTag" :disabled="props.disabled" :class="buttonClasses" data-testid="button">
    <slot name="prepend"  v-if="$slots.prepend" />
    <slot />
    <slot name="append" v-if="$slots.append" />
  </component>
</template>
