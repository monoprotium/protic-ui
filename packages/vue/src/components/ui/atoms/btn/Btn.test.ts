import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import PrtBtn from './PrtBtn.vue';
import { PrtBtnSize, PrtBtnVariant } from './types';

// Default props
const defaultProps = {
  size: PrtBtnSize.base,
  variant: PrtBtnVariant.solid,
  disabled: false,
  square: false,
  tag: 'button',
  loading: false,
  fullWidth: false,
  color: 'bg-el-7',
};

describe('PrtBtn', () => {
  // Initial State
  it('matches the initial state snapshot', () => {
    const wrapper = mount(PrtBtn, {
      props: defaultProps,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  // Sizes Classes
  it.each(Object.values(PrtBtnSize))('renders the correct size for size variant %s', (size) => {
    const wrapper = mount(PrtBtn, {
      props: { ...defaultProps, size },
    });

    const expectedClass = {
      [PrtBtnSize.sm]: 'text-sm',
      [PrtBtnSize.base]:  'text-base', 
      [PrtBtnSize.lg]: 'text-base', 
    };
    expect(wrapper.classes()).toContain(expectedClass[size]);
  });

  // Variant Classes
  it.each(Object.values(PrtBtnVariant))('renders the correct styling for variant %s', (variant) => {
    const wrapper = mount(PrtBtn, {
      props: { ...defaultProps, variant },
    });
  
    // Check for the correct classes applied to the component
    const expectedClasses = {
      [PrtBtnVariant.solid]: ['text-white', 'bg-el-7', 'hover:bg-opacity-90', 'focus:ring-2', 'focus:ring-offset-2', 'focus:ring-el-7', 'focus:ring-opacity-50'],
      [PrtBtnVariant.outline]: ['bg-transparent', 'border-el-7', 'hover:text-white', 'text-el-7', 'hover:bg-el-7', 'focus:ring-2', 'focus:ring-offset-2', 'focus:ring-el-7', 'focus:ring-opacity-50'],
      [PrtBtnVariant.ghost]: ['text-el-7', 'bg-transparent', 'hover:bg-el-7', 'hover:text-white', 'focus:ring-2', 'focus:ring-offset-2', 'focus:ring-el-7', 'focus:ring-opacity-50'],
    };
  
    expectedClasses[variant].forEach((expectedClass) => {
      expect(wrapper.classes()).toContain(expectedClass);
    });
  });
  // Disabled State
  it('renders in a disabled state when disabled prop is true', () => {
    const wrapper = mount(PrtBtn, {
      props: { ...defaultProps, disabled: true },
    });
    expect(wrapper.attributes('disabled')).toBeDefined();
  });

  // Square Button
  it('renders as a square button when square prop is true', () => {
    const wrapper = mount(PrtBtn, {
      props: { ...defaultProps, square: true },
    });

    const expectedClasses = ['p-2', 'min-w-10.5'];

    expectedClasses.forEach((expectedClass) => {
      expect(wrapper.classes()).toContain(expectedClass);
    });
  });



  // Prepend Slot
  it('displays the button with the specified prepend slot', () => {
    const wrapper = mount(PrtBtn, {
      props: defaultProps,
      slots: {
        prepend: '<span>Prepend</span>',
      },
    });
    expect(wrapper.html()).toContain('Prepend');
  });

  // Append Slot
  it('displays the button with the specified append slot', () => {
    const wrapper = mount(PrtBtn, {
      props: defaultProps,
      slots: {
        append: '<span>Append</span>',
      },
    });
    expect(wrapper.html()).toContain('Append');
  });

  // Prepend, Append, and Default Content
  it('displays the button with prepend, append, and default slots', () => {
    const wrapper = mount(PrtBtn, {
      props: defaultProps,
      slots: {
        prepend: '<span>Prepend</span>',
        append: '<span>Append</span>',
        default: '<span>Content</span>', 
      },
    });

    expect(wrapper.html()).toContain('<span>Prepend</span>');
    expect(wrapper.html()).toContain('<span>Append</span>');
    expect(wrapper.html()).toContain('<span>Content</span>');
  });

});
