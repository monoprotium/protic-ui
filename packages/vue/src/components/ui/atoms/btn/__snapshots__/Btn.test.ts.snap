// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrtBtn > matches the initial state snapshot 1`] = `
"<button type=\\"button\\" class=\\"inline-flex items-center justify-center font-medium transition-colors duration-150 rounded text-center  text-base py-2 px-4    text-white bg-el-7 hover:bg-opacity-90 focus:ring-2 focus:ring-offset-2 focus:ring-el-7 focus:ring-opacity-50\\" data-testid=\\"button\\">
  <!--v-if-->
  <!--v-if-->
</button>"
`;
